lxd/project/permissions.go

```diff
@@ -1514,3 +1514,107 @@ func AllowSnapshotCreation(p *api.Project) error {
 
 	return nil
 }
+
+// GetRestrictedClusterGroups returns a slice of restricted cluster groups for the given project.
+func GetRestrictedClusterGroups(p *api.Project) []string {
+	return shared.SplitNTrimSpace(p.Config["restricted.cluster.groups"], ",", -1, true)
+}
+
+// AllowClusterMember returns nil if the given project is allowed to use the cluster member.
+func AllowClusterMember(p *api.Project, member *db.NodeInfo) error {
+	clusterGroupsAllowed := GetRestrictedClusterGroups(p)
+
+	if shared.IsTrue(p.Config["restricted"]) && len(clusterGroupsAllowed) > 0 {
+		for _, memberGroupName := range member.Groups {
+			if shared.StringInSlice(memberGroupName, clusterGroupsAllowed) {
+				return nil
+			}
+		}
+
+		return fmt.Errorf("Project isn't allowed to use this cluster member")
+	}
+
+	return nil
+}
+
+// AllowClusterGroup returns nil if the given project is allowed to use the cluster group.
+func AllowClusterGroup(p *api.Project, groupName string) error {
+	clusterGroupsAllowed := GetRestrictedClusterGroups(p)
+
+	// Skip the check if the project is not restricted
+	if shared.IsFalseOrEmpty(p.Config["restricted"]) {
+		return nil
+	}
+
+	if len(clusterGroupsAllowed) > 0 && !shared.StringInSlice(groupName, clusterGroupsAllowed) {
+		return fmt.Errorf("Project isn't allowed to use this cluster group")
+	}
+
+	return nil
+}
+
+// CheckTargetMember checks if the given member is allowed for the project.
+// If nil it also returns the resolved target member node information.
+func CheckTargetMember(ctx context.Context, tx *db.ClusterTx, p *api.Project, targetMember string) (*db.NodeInfo, error) {
+	// Find target member.
+	member, err := tx.GetNodeByName(ctx, targetMember)
+	if err != nil {
+		return nil, api.StatusErrorf(http.StatusNotFound, "Cluster member not found")
```

Comment:
I know we are getting rid of this call now anyway, but for future reference we shouldn't do this sort of error handling.
DB queries can fail for a number of different reasons and so masking all errors as a "cluster member not found error" is not that helpful to identify the actual cause of the problem. Also this can be problematic because the caller may want to take different action depending on whether there was a not found error or an unexpected error.

Also the `GetNodeByName` function itself returns `api.StatusErrorf(http.StatusNotFound, "Cluster member not found")` if the member wasn't found so there's no need to mask it here.

Reply:
Just to get it right, you mean putting additional information in the error message without losing the actual cause:

```
api.StatusErrorf(http.StatusNotFound, "Cluster member not found: %q", err.Error())
```

Reply:
No I mean just returning `err` here or, if you want to add context, doing something like `return fmt.Errorf("Failed loading cluster member: %w", err)` that way using `%w` means the `api.StatusError` type will still be propagated upwards and will then influence the HTTP response code and any upstream callers of `api.StatusErrorCheck()`. 

See https://go.dev/blog/go1.13-errors

So the two key points are:

- Don't mask errors with different meanings. E.g. don't say "Cluster member not found" unless you know that is the error (it maybe the network is down, for example, but that is not a cluster member not found error). The best place for detecting not found errors and returning them as such is in the actual DB function, not an upstream caller. So it looks like a code smell to me when I see it further up.
- Use %w for wrapping errors, not %v, %q or %s.

Reply:
You only need to return an `api.StatusError` if you want to indicate a specific error code. Otherwise just return a normal error and the API handlers will convert that to a 500 HTTP error.